### PR TITLE
Fix potential memory leak caused by zookeeper exist calling

### PR DESCRIPTION
**Pull request by hustnn**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix potential memory leak caused by zookeeper exists watch.

Detailed description / Documentation draft:
Closes #14626

### CODE REVIEW

**Comment by hustnn** [comment]
@alexey-milovidov Can you help to look at this when you are free? Thanks.

File: src/Common/ZooKeeper/ZooKeeperImpl.cpp [hunk 1]
```diff
@@ -1278,6 +1276,22 @@ void ZooKeeper::receiveEvent()
             response->removeRootPath(root_path);
         }
 
+        /// Find from operations
+        if (request_info.watch)
+        {
+            bool add_watch = false;
+            if (request_info.isExists)
```
Comment (alexey-milovidov) [inline src/Common/ZooKeeper/ZooKeeperImpl.cpp:1283]: The idea is Ok.

Maybe dynamic_cast will be better than adding a field?

Comment (alexey-milovidov) [reply src/Common/ZooKeeper/ZooKeeperImpl.cpp:1283]: Also please add some explanation in comments.

Comment (alexey-milovidov) [reply src/Common/ZooKeeper/ZooKeeperImpl.cpp:1283]: getOpNum may also suffice.

Comment (hustnn) [reply src/Common/ZooKeeper/ZooKeeperImpl.cpp]: I changed to use the getOpNum. Seems the test case `test_chroot_with_same_root` failed. It should be not related to this change, I need to look at why.

Comment (alexey-milovidov) [reply src/Common/ZooKeeper/ZooKeeperImpl.cpp]: There were no failures of this test before.

Comment (hustnn) [reply src/Common/ZooKeeper/ZooKeeperImpl.cpp]: Yes, this error occurs after this PR. I fixed your comment first, I mean this test failure should still have, it is caused by other place in this PR.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by hustnn** [comment]
@alexey-milovidov The change is still not correct. --!
 I changed to [WIP] status first.

**Comment by hustnn** [comment]
Some tests which should be related to this change also fail. A little strange.

**Comment by hustnn** [comment]
rebuild

File: src/Common/ZooKeeper/ZooKeeperImpl.cpp [hunk 2]
```diff
@@ -1278,6 +1288,30 @@ void ZooKeeper::receiveEvent()
             response->removeRootPath(root_path);
         }
 
+        /// Instead of setting the watch in sendEvent, set it in receiveEvent becuase need to check the response.
```
Comment (alexey-milovidov) [inline src/Common/ZooKeeper/ZooKeeperImpl.cpp:1291]: Ok. It looks correct *under assumption that the responses from ZooKeeper are received in order*.

For example, if we have received a response for a command where a watch was set, then we cannot receive the corresponding watch event earlier.

Are you sure that this assumption is correct?

**Comment by alexey-milovidov** [comment]
https://zookeeper.apache.org/doc/r3.6.1/zookeeperProgrammers.html#ch_zkWatches

> With regard to watches, ZooKeeper maintains these guarantees:
- Watches are ordered with respect to other events, other watches, and asynchronous replies.

Ok.

**Comment by hustnn** [comment]
> https://zookeeper.apache.org/doc/r3.6.1/zookeeperProgrammers.html#ch_zkWatches
> 
> > With regard to watches, ZooKeeper maintains these guarantees:
> > 
> > * Watches are ordered with respect to other events, other watches, and asynchronous replies.
> 
> Ok.

@alexey-milovidov Should we add some validate logic in client to make sure the requests are processed in order?

**Comment by alexey-milovidov** [comment]
Not necessary... I think it will be too much burden.

**Comment by amosbird** [comment]
The metric `ZooKeeperWatch` never decrements after this PR, and the speed of zxid consumption increased drastically.

**Pull request merged**